### PR TITLE
[top,dv,rv_dm] ndm reset request test

### PR DESCRIPTION
**Pull request by jdonjdon**
NDM reset request test
This test does csr write to  lc / sys reset regions. Then it toggles ndm reset.
After toggling reset, it reads back the same csrs to check
 the csr under lc reset region doesn't change the value 
while the csr under sys reset region has reset value.

Signed-off-by: Jaedon Kim <jdonjdon@google.com>

### CODE REVIEW

**Comment by tjaychen** [comment]
i just triggered an fpga re-run, should we work on merging this PR now?

**Comment by jdonjdon** [comment]
> i just triggered an fpga re-run, should we work on merging this PR now?

yep, will do soon.

**Comment by jdonjdon** [comment]
@tjaychen I had to re submit to clean-up some debug codes.

File: sw/device/tests/sim_dv/rv_dm_ndm_reset_req.c [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+
+static test_register_t kReg[] = {
+    [OTP_CTRL] =
+        {
+            .name = "OTP_CTRL",
+            .dif_handle = &otp_ctrl,
+            .write_val = 0x06092022,
+            .exp_read_val = 0x06092022,
+        },
+    [PINMUX] =
```
Comment (tjaychen) [inline sw/device/tests/sim_dv/rv_dm_ndm_reset_req.c:107]: is the pinmux one still right? I thought we changed the reset domain of the pinmux so the wouldn't the `exp_read_val` go back to default? Might need a rebase for this one.

File: sw/device/tests/sim_dv/rv_dm_ndm_reset_req.c [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+        {
+            .name = "PINMUX",
+            .write_val = 0x44,
+            .exp_read_val = 0x44,
+            .dif_handle = &pinmux,
+        },
+    [SRAM_CTRL] =
+        {
```
Comment (tjaychen) [inline sw/device/tests/sim_dv/rv_dm_ndm_reset_req.c:115]: should `sram_ctrl` expected read value return to default also?

File: sw/device/tests/sim_dv/rv_dm_ndm_reset_req.c [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+  OTP_CTRL,
+  PINMUX,
+  SRAM_CTRL,
+  ADC_CTRL,
+  SYSRST_CTRL,
+  KEYMGR,
+  FLASH_CTRL,
+};
+
+static test_register_t kReg[] = {
```
Comment (tjaychen) [inline sw/device/tests/sim_dv/rv_dm_ndm_reset_req.c:99]: let me know if i understood it correctly, from what i remember, i thought the `exp_rd_val` was supposed to be the value after an ndm reset, if so, a couple below look like they might be wrong.

File: sw/device/tests/sim_dv/rv_dm_ndm_reset_req.c (hunk 1 above)
Comment (jdonjdon) [reply sw/device/tests/sim_dv/rv_dm_ndm_reset_req.c:107]: > is the pinmux one still right? I thought we changed the reset domain of the pinmux so the wouldn't the `exp_read_val` go back to default? Might need a rebase for this one.
There are LC group and SYS group csr.
PINMUX is moved to LC group and FLASH is moved to SYS group.

File: sw/device/tests/sim_dv/rv_dm_ndm_reset_req.c (hunk 2 above)
Comment (jdonjdon) [reply sw/device/tests/sim_dv/rv_dm_ndm_reset_req.c:115]: this is retention portion. I will change name to SRAM_CTRL_RET.

Comment (tjaychen) [reply sw/device/tests/sim_dv/rv_dm_ndm_reset_req.c:115]: hmm.. but isn't retention sram using [sys reset](https://github.com/lowRISC/opentitan/blob/master/hw/top_earlgrey/data/top_earlgrey.hjson#L517) also?

File: sw/device/tests/sim_dv/rv_dm_ndm_reset_req.c (hunk 1 above)
Comment (tjaychen) [reply sw/device/tests/sim_dv/rv_dm_ndm_reset_req.c:107]: o that's right, i mixed up pinmux in my head.

File: sw/device/tests/sim_dv/rv_dm_ndm_reset_req.c (hunk 2 above)
Comment (tjaychen) [reply sw/device/tests/sim_dv/rv_dm_ndm_reset_req.c:115]: OOOOO. thank you that makes sense.

File: sw/device/tests/sim_dv/rv_dm_ndm_reset_req.c [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+#include "adc_ctrl_regs.h"
+#include "flash_ctrl_regs.h"
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+#include "keymgr_regs.h"
+#include "otp_ctrl_regs.h"
+#include "pinmux_regs.h"
+#include "sysrst_ctrl_regs.h"
+/*
+   RV_DM NDM RESET REQUEST TEST
+   In top_earlgrey, csr can be devided 3 groups as below.
```
Comment (a-will) [inline sw/device/tests/sim_dv/rv_dm_ndm_reset_req.c:26]: nit:

```suggestion
   In top_earlgrey, the CSRs can be divided into 3 groups as below.
```

File: sw/device/tests/sim_dv/rv_dm_ndm_reset_req.c [hunk 5]
```diff
@@ -0,0 +1,262 @@
...
+                      kReg[KEYMGR].write_val);
+}
+
+static uint32_t read_test_reg(uint32_t device_idx, bool is_initial) {
+  mmio_region_t addr;
+  ptrdiff_t offset;
+  switch (device_idx) {
+    case OTP_CTRL:
+      addr = mmio_region_from_addr(TOP_EARLGREY_OTP_CTRL_CORE_BASE_ADDR);
+      offset = OTP_CTRL_DIRECT_ACCESS_WDATA_0_REG_OFFSET;
```
Comment (a-will) [inline sw/device/tests/sim_dv/rv_dm_ndm_reset_req.c:180]: (just a question / suggestion) Would it make sense to throw these in the `test_register_t` table above? It seems like that would keep the values and offsets together, improving readability.

But it's also fine as-is.

**Review by a-will — APPROVED** [review]
LGTM for software, with a minor typo nit + a question.

File: sw/device/tests/sim_dv/rv_dm_ndm_reset_req.c (hunk 4 above)
Comment (jdonjdon) [reply sw/device/tests/sim_dv/rv_dm_ndm_reset_req.c]: fixed

**Pull request merged**